lib/auth/autoupdate_rollout.go

```diff
@@ -0,0 +1,174 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"math/rand/v2"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/constants"
+	autoupdatev1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/autoupdate/v1"
+	"github.com/gravitational/teleport/lib/inventory"
+)
+
+// SampleAgentsFromAutoUpdateGroup iterates over every handle in the inventory to
+// build a random sample of agents belonging to a given group.
+// The main use-case for this function is to pick canaries that can be updated.
+func (a *Server) SampleAgentsFromAutoUpdateGroup(ctx context.Context, groupName string, sampleSize int, groups []string) ([]*autoupdatev1pb.Canary, error) {
+	if len(groups) == 0 {
+		return nil, trace.BadParameter("no groups specified")
+	}
+	isCatchAll := groupName == groups[len(groups)-1]
+	var groupSet map[string]struct{}
+
+	// Small optimization, we only need to build the groupSet if we are sampling the catch-all group
+	if isCatchAll {
+		groupSet = make(map[string]struct{})
+		for _, group := range groups {
+			groupSet[group] = struct{}{}
+		}
+	}
+
+	filter := func(handle inventory.UpstreamHandle) bool {
+		ok, _ := filterHandler(handle, a.clock.Now())
+		if !ok {
+			return false
+		}
+
+		// If this is not the catch-all group, we can only check if the agent group is the right one.
+		if !isCatchAll {
+			// No need to check for UpdaterInfo being nil, it would have been filtered
+			// out by filterHandler().
+			return handle.Hello().UpdaterInfo.UpdateGroup == groupName
+		}
+		// This is the catch-call group, it matches agents from every group not in groups.
+		_, ok = groupSet[groupName]
+		// If the agent group is not in the group list, it falls into the catch-all.
+		return !ok
+	}
+
+	sampler := newHandlerSampler(sampleSize, filter)
+
+	a.inventory.UniqueHandles(sampler.visit)
+
+	sampled := sampler.Sampled()
+	canaries := make([]*autoupdatev1pb.Canary, len(sampled))
```

Comment:
*nit*: Might use capacity pre-allocation + append here, to match other logic, but no strong opinion.